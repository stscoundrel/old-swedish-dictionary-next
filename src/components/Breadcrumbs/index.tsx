import { BreadcrumbRequest, getBreadcrumbs } from 'lib/utils/breadcrumbs'
import { getBreadcrumbsSchema } from 'lib/services/schema'
import Link from 'next/link'
import { DictionaryEntry } from 'lib/models/dictionary'
import { AlphabetLetter } from 'lib/services/dictionary'
import styles from './Breadcrumbs.module.scss'

interface BreadcrumbProps{
  word: DictionaryEntry | null,
  letter: AlphabetLetter | null
}

export default function Breadcrumbs({ word, letter }: BreadcrumbProps) {
  const getBreadCrumbsRequest = (): BreadcrumbRequest => {
    const request: BreadcrumbRequest = {
      word: null,
      letter: null,
    }

    if (word) {
      request.word = word.headword
    }

    if (letter) {
      request.letter = letter.letter
    }

    return request
  }

  const breadcrumbs = getBreadcrumbs(getBreadCrumbsRequest())
  const schema = getBreadcrumbsSchema(breadcrumbs)

  return (
    <nav className={styles.section}>
      <div className="container">
        {breadcrumbs.map(({ label, url }) => (
          <Link key={url} href={url} className={styles.link} prefetch={false}>
            {label}
          </Link>
        ))}

        <script type='application/ld+json' dangerouslySetInnerHTML={ { __html: schema } }/>
      </div>
    </nav>
  )
}
