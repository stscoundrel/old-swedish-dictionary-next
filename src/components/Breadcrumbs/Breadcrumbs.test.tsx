import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import Breadcrumbs from './index'

describe('Breadcrumbs component', () => {
  const word = {
    headword: 'afköra',
    partOfSpeech: ['vb'],
    grammaticalAspect: 'v.',
    definitions: [
      ' , vräka. hafwi wald han afköra  GS 43 (1416?). af ty ahren (för areno) han landboen afkörer ib.',
    ],
    alternativeForms: ['-bortdrifva'],
    slug: 'afkora',
  }

  const letter = {
    letter: 'a',
    slug: 'a',
  }

  test('Default page: does not crach', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Breadcrumbs word={null} letter={null} />)
  })

  test('Default page: matches snapshot', () => {
    const tree = renderer.create(<Breadcrumbs word={null} letter={null} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Default page: renders correct amount of breadcrumbs', () => {
    const tree = renderer.create(<Breadcrumbs word={null} letter={null} />)
    const { root } = tree

    expect(root.findAllByType('a').length).toEqual(1)
  })

  test('Letter page: does not crach', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Breadcrumbs word={null} letter={letter} />)
  })

  test('Letter page: matches snapshot', () => {
    const tree = renderer.create(
      <Breadcrumbs word={null} letter={letter} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Letter page: renders correct amount of breadcrumbs', () => {
    const tree = renderer.create(<Breadcrumbs word={null} letter={letter} />)
    const { root } = tree

    expect(root.findAllByType('a').length).toEqual(2)
  })

  test('Word page: does not crach', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Breadcrumbs word={word} letter={letter} />)
  })

  test('Word page: matches snapshot', () => {
    const tree = renderer.create(<Breadcrumbs word={word} letter={letter} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Word page: renders correct amount of breadcrumbs', () => {
    const tree = renderer.create(<Breadcrumbs word={word} letter={letter} />)
    const { root } = tree

    expect(root.findAllByType('a').length).toEqual(3)
  })
})
