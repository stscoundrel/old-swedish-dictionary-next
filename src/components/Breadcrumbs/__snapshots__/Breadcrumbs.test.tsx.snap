// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs component Default page: matches snapshot 1`] = `
<nav
  className="section"
>
  <div
    className="container"
  >
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Old Swedish Dictionary
    </a>
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Old Swedish Dictionary\\",\\"item\\":\\"undefined/\\"}]}",
        }
      }
      type="application/ld+json"
    />
  </div>
</nav>
`;

exports[`Breadcrumbs component Letter page: matches snapshot 1`] = `
<nav
  className="section"
>
  <div
    className="container"
  >
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Old Swedish Dictionary
    </a>
    <a
      href="/letter/a"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Letter A
    </a>
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Old Swedish Dictionary\\",\\"item\\":\\"undefined/\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"name\\":\\"Letter A\\",\\"item\\":\\"undefined/letter/a\\"}]}",
        }
      }
      type="application/ld+json"
    />
  </div>
</nav>
`;

exports[`Breadcrumbs component Word page: matches snapshot 1`] = `
<nav
  className="section"
>
  <div
    className="container"
  >
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Old Swedish Dictionary
    </a>
    <a
      href="/letter/a"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Letter A
    </a>
    <a
      href="/word/afkora"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      Afköra
    </a>
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Old Swedish Dictionary\\",\\"item\\":\\"undefined/\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"name\\":\\"Letter A\\",\\"item\\":\\"undefined/letter/a\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":3,\\"name\\":\\"Afköra\\",\\"item\\":\\"undefined/word/afkora\\"}]}",
        }
      }
      type="application/ld+json"
    />
  </div>
</nav>
`;
